kyuubi-server/src/main/scala/org/apache/kyuubi/engine/EngineRef.scala

```diff
@@ -126,11 +130,33 @@ private[kyuubi] class EngineRef private(conf: KyuubiConf, user: String, sessionI
   }
 
   private def create(zkClient: CuratorFramework): (String, Int) = tryWithLock(zkClient) {
-    // TODO: improve this after support engine pool. (KYUUBI #918)
-    var engineRef = getServerHost(zkClient, engineSpace)
-    // Get the engine address ahead if another process has succeeded
-    if (engineRef.nonEmpty) return engineRef.get
 
+    // USER share level support engine pool
+    if (shareLevel.equals(USER)) {
+
+      // create engine space
+      createEngineSpaceIfNotExists(zkClient, engineSpace, conf.get(HA_ZK_ENGINE_POOL_SIZE))
+
+      // create engine if not reach pool size
```

Comment:
How can we reach to pool size？

Reply:
When creating engine, will check `pool.size` and `engineNum`.

_`pool size`_: read from engine space.
_`engine num`_: children of engine space.


![image](https://user-images.githubusercontent.com/86483005/130398851-5048ae25-d0a3-4041-a25d-5315afe1b10e.png)

Reply:
can you add a test to verify this？

Reply:
Updated `KyuubiOperationPerUserSuite`, I am not sure if the coding is elegant.

Reply:
So, the pool actually makes connect 2 a very worse case
1)  it always needs to create a new engine even there is an available engine there. the time cost is much higher than just reusing.
2)  it will fail if it fails to create a new engine. The failure risk is much higher to create more engines.
3) if we can fall back to reusing after failures in 2), the time cost is inevitable.

If the failure happens, connection 3 seems very like to repeat what connection 2 does.

Reply:
Thanks for raising concerns:

> it always needs to create a new engine even there is an available engine there

This is the characteristic of the auto-mechanism, we can provide option to skip create a new engine.

> if we can fall back to reusing after failures in 2), the time cost is inevitable.

This work as `subDomain`. When using `subDomain` and encountering failure, time-consuming is also unavoidable.

**Improve**

- provide option to skip create a new engine
- fall back to reusing after creating failure

Reply:
> This is the characteristic of the auto-mechanism, we can provide option to skip create a new engine.

It makes the feature itself even more unfriendly.

> This work as subDomain. When using subDomain and encountering failure, time-consuming is also unavoidable.

Not really the same. the original intention of a subdomain is to provide isolation but creating multiple engines under a subdomain is for speed and power. But now, it does bring extra creation overhead for a lot of cases

Reply:
For most scene, we use sparksql as offline analysis, `engine pool` is useful for us. Ok, if user very case about the time cost, keep `kyuubi.ha.engine.pool.size=1`, the kyuubi server will work like before.

Reply:
Not exactly too, you add a zookeeper distribute lock for all connections, this is much heavier than before

Reply:
we are talking about the engine pool itself, for what I am trying to improve based on your current implementation. I am not sure if my concerns can be ignored or persuade like that.